src/esp/gfx/PbrDrawable.h

```diff
@@ -5,15 +5,24 @@
 #ifndef ESP_GFX_PBRDRAWABLE_H_
 #define ESP_GFX_PBRDRAWABLE_H_
 
+#include <Corrade/Containers/Optional.h>
+
 #include "esp/gfx/Drawable.h"
 #include "esp/gfx/PbrImageBasedLighting.h"
 #include "esp/gfx/PbrShader.h"
 #include "esp/gfx/ShaderManager.h"
+#include "esp/gfx/ShadowMapManager.h"
 namespace esp {
 namespace gfx {
 
 class PbrDrawable : public Drawable {
  public:
+  struct ShadowData {
+    // all the terms must be set by the user
+    ShadowMapManager* shadowMapManger = nullptr;
+    ShadowMapKeys* shadowMapKeys = nullptr;
+  };
```

Comment:
Not sure about the usefulness / usability of this struct -- it has pointers on which you have to assert for non-null and then you only use it in a single function, `setShadowData()`, *and then* you save it into an optional, which adds another level of potential null-ness on which the code has to branch.

Since the two members are of a different type, this is not the case where one could mix up the parameters, so why not directly have `setShadowData(ShadowMapManager&, ShadowMapKeys&, PbrShader::Flag)`? Easier for the user to supply as there's no need to fill a temporary struct. You can then save those to the class as two separate pointers, no need for a struct there either -- it's internal state of the class, so the public interface already ensures that those are either both set, or when `setShadowData()` wasn't called, then they are both null.

Reply:
Ha, it totally makes sense. Let me try your suggestion.
You know what, the reason I had this struct here is that it is used to contain more values than its current shape. e.g., `lightNearPlane`, `lightFarPlane`. 
I removed them in this diff because I abandoned `traditional shadow cube map + Percentage Closer Filtering (PCF)` -- another way to render shadows. I use VSM here.